src/editors.ts

```diff
@@ -0,0 +1,369 @@
+    return editor;
+  }
+
+  active(): boolean {
+    return (
+      !!document.activeElement &&
+      !!document.activeElement!.closest(".ace_editor")!
+    );
+  }
+
+  getEditorState(): EditorState {
+    const editor = this.editor();
+    const source = editor.getValue();
+    const { row, column } = editor.getCursorPosition();
+    let modeParts = editor.session.getMode()["$id"].split("/");
+    const mode = modeParts[modeParts.length - 1];
+    return {
+      source,
+      cursor: this.cursorFromRowAndColumn(source, row, column),
+      filename: this.filenameFromLanguage(mode),
+      available: true,
+    };
+  }
+
+  setSelection(cursor: number, cursorEnd: number) {
+    const editor = this.editor();
+    const source = editor.getValue();
+    const positionStart = this.rowAndColumnFromCursor(source, cursor);
+    const positionEnd = this.rowAndColumnFromCursor(source, cursorEnd);
+    editor.session.selection.setRange(
+      positionStart.row,
+      positionStart.column,
+      positionEnd.row,
+      positionEnd.column
+    );
+  }
+
+  setSourceAndCursor(source: string, cursor: number) {
+    const editor = this.editor();
+    editor.session.setValue(source);
+    const { row, column } = this.rowAndColumnFromCursor(source, cursor);
+    editor.session.selection.moveCursorTo(row, column);
+  }
+
+  redo() { }
+
+  undo() { }
+}
+
+class CodeMirror extends Editor {
+  private editor(): any {
+    const codemirror = document.activeElement!.closest(".CodeMirror")!;
+    return (codemirror as any).CodeMirror;
+  }
+
+  active(): boolean {
+    return (
+      !!document.activeElement &&
+      !!document.activeElement.closest(".CodeMirror")
+    );
+  }
+
+  getEditorState(): EditorState {
+    const editor = this.editor();
+    const source = editor.getValue();
+    const { line, ch } = editor.getCursor();
+    let mode = editor.options.mode;
+    if (mode && typeof mode !== "string" && mode.name) {
+      mode = mode.name;
+    }
+    const filename = this.filenameFromLanguage(mode);
+    return {
+      source,
+      cursor: this.cursorFromRowAndColumn(source, line, ch),
+      filename: filename,
+      available: true,
+    };
+  }
+
+  setSelection(cursor: number, cursorEnd: number) {
+    const editor = this.editor();
+    const positionStart = this.rowAndColumnFromCursor(
+      editor.getValue(),
+      cursor
+    );
+    const positionEnd = this.rowAndColumnFromCursor(
+      editor.getValue(),
+      cursorEnd
+    );
+    editor.setSelection(
+      { line: positionStart.row, ch: positionStart.column, sticky: null },
+      { line: positionEnd.row, ch: positionEnd.column, sticky: null }
+    );
+  }
+
+  setSourceAndCursor(source: string, cursor: number) {
+    const editor = this.editor();
+    editor.setValue(source);
+    const { row, column } = this.rowAndColumnFromCursor(source, cursor);
+    editor.setCursor(row, column);
+  }
+
+  redo() {
+    const editor = this.editor();
+    editor.redo();
```

Comment:
can just inline this to `this.editor().redo()` if we're only going to use the `editor` object once.